File: driver-scala/src/it/scala/org/mongodb/scala/syncadapter/SyncMongoClient.scala
package org.mongodb.scala.syncadapter

import scala.collection.JavaConverters._
import com.mongodb.ClientSessionOptions
import com.mongodb.client.{ ClientSession, MongoClient => JMongoClient, MongoDatabase => JMongoDatabase }
import org.bson.Document
import org.bson.conversions.Bson
import org.mongodb.scala.MongoClient
import org.mongodb.scala.bson.DefaultHelper.DefaultsTo

import scala.concurrent.Await
import scala.concurrent.duration.Duration
import scala.reflect.ClassTag

case class SyncMongoClient(wrapped: MongoClient) extends JMongoClient {

  override def getDatabase(databaseName: String): JMongoDatabase =
    new SyncMongoDatabase(wrapped.getDatabase(databaseName))

  override def startSession: ClientSession =
    SyncClientSession(Await.result(wrapped.startSession().head(), Duration(10, "second")), this)

  override def startSession(options: ClientSessionOptions): ClientSession =
    SyncClientSession(Await.result(wrapped.startSession(options).head(), Duration(10, "second")), this)

  override def close(): Unit = wrapped.close()

  override def listDatabaseNames = throw new UnsupportedOperationException

  override def listDatabaseNames(clientSession: ClientSession) = throw new UnsupportedOperationException

  override def listDatabases = new SyncListDatabasesIterable[Document](wrapped.listDatabases[Document]())

  override def listDatabases(clientSession: ClientSession) = throw new UnsupportedOperationException

  override def listDatabases[TResult](resultClass: Class[TResult]) =
    new SyncListDatabasesIterable[TResult](
      wrapped.listDatabases[TResult]()(
        DefaultsTo.overrideDefault[TResult, org.mongodb.scala.Document],
        ClassTag(resultClass)
      )
    )

  override def listDatabases[TResult](clientSession: ClientSession, resultClass: Class[TResult]) =
    throw new UnsupportedOperationException

  override def watch = new SyncChangeStreamIterable[Document](wrapped.watch[Document]())

  override def watch[TResult](resultClass: Class[TResult]) =
    new SyncChangeStreamIterable[TResult](
      wrapped.watch[TResult]()(DefaultsTo.overrideDefault[TResult, org.mongodb.scala.Document], ClassTag(resultClass))
    )

  override def watch(pipeline: java.util.List[_ <: Bson]) =
    new SyncChangeStreamIterable[Document](wrapped.watch[Document](pipeline.asScala.toSeq))

  override def watch[TResult](pipeline: java.util.List[_ <: Bson], resultClass: Class[TResult]) =
    new SyncChangeStreamIterable[TResult](
      wrapped.watch[TResult](pipeline.asScala.toSeq)(
        DefaultsTo.overrideDefault[TResult, org.mongodb.scala.Document],
        ClassTag(resultClass)
      )
    )

  override def watch(clientSession: ClientSession) =
    new SyncChangeStreamIterable[Document](wrapped.watch[Document](unwrap(clientSession)))

  override def watch[TResult](clientSession: ClientSession, resultClass: Class[TResult]) =
    new SyncChangeStreamIterable[TResult](
      wrapped.watch(unwrap(clientSession))(
        DefaultsTo.overrideDefault[TResult, org.mongodb.scala.Document],
        ClassTag(resultClass)
      )
    )

  override def watch(clientSession: ClientSession, pipeline: java.util.List[_ <: Bson]) =
    new SyncChangeStreamIterable[Document](wrapped.watch[Document](unwrap(clientSession), pipeline.asScala.toSeq))

  override def watch[TResult](
      clientSession: ClientSession,
      pipeline: java.util.List[_ <: Bson],
      resultClass: Class[TResult]
  ) =
    new SyncChangeStreamIterable[TResult](
      wrapped.watch[TResult](unwrap(clientSession), pipeline.asScala.toSeq)(
        DefaultsTo.overrideDefault[TResult, org.mongodb.scala.Document],
        ClassTag(resultClass)
      )
    )

  override def getClusterDescription = throw new UnsupportedOperationException

  private def unwrap(clientSession: ClientSession): org.mongodb.scala.ClientSession =
    clientSession.asInstanceOf[SyncClientSession].wrapped
}

File: driver-kotlin-sync/src/integration/kotlin/com/mongodb/kotlin/client/syncadapter/SyncMongoClient.kt
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mongodb.kotlin.client.syncadapter

import com.mongodb.ClientSessionOptions
import com.mongodb.client.ChangeStreamIterable
import com.mongodb.client.ClientSession
import com.mongodb.client.ListDatabasesIterable
import com.mongodb.client.MongoClient as JMongoClient
import com.mongodb.client.MongoDatabase
import com.mongodb.client.MongoIterable
import com.mongodb.connection.ClusterDescription
import com.mongodb.kotlin.client.MongoClient
import org.bson.Document
import org.bson.conversions.Bson

internal class SyncMongoClient(val wrapped: MongoClient) : JMongoClient {
    override fun close(): Unit = wrapped.close()

    override fun getDatabase(databaseName: String): MongoDatabase = SyncMongoDatabase(wrapped.getDatabase(databaseName))

    override fun startSession(): ClientSession = SyncClientSession(wrapped.startSession(), this)

    override fun startSession(options: ClientSessionOptions): ClientSession =
        SyncClientSession(wrapped.startSession(options), this)

    override fun listDatabaseNames(): MongoIterable<String> = SyncMongoIterable(wrapped.listDatabaseNames())

    override fun listDatabaseNames(clientSession: ClientSession): MongoIterable<String> =
        SyncMongoIterable(wrapped.listDatabaseNames(clientSession.unwrapped()))

    override fun listDatabases(): ListDatabasesIterable<Document> = SyncListDatabasesIterable(wrapped.listDatabases())

    override fun listDatabases(clientSession: ClientSession): ListDatabasesIterable<Document> =
        SyncListDatabasesIterable(wrapped.listDatabases(clientSession.unwrapped()))

    override fun <T : Any> listDatabases(resultClass: Class<T>): ListDatabasesIterable<T> =
        SyncListDatabasesIterable(wrapped.listDatabases(resultClass))

    override fun <T : Any> listDatabases(
        clientSession: ClientSession,
        resultClass: Class<T>
    ): ListDatabasesIterable<T> =
        SyncListDatabasesIterable(wrapped.listDatabases(clientSession.unwrapped(), resultClass))

    override fun watch(): ChangeStreamIterable<Document> = SyncChangeStreamIterable(wrapped.watch())

    override fun <T : Any> watch(resultClass: Class<T>): ChangeStreamIterable<T> =
        SyncChangeStreamIterable(wrapped.watch(resultClass = resultClass))

    override fun watch(pipeline: MutableList<out Bson>): ChangeStreamIterable<Document> =
        SyncChangeStreamIterable(wrapped.watch(pipeline))

    override fun <T : Any> watch(pipeline: MutableList<out Bson>, resultClass: Class<T>): ChangeStreamIterable<T> =
        SyncChangeStreamIterable(wrapped.watch(pipeline, resultClass))

    override fun watch(clientSession: ClientSession): ChangeStreamIterable<Document> =
        SyncChangeStreamIterable(wrapped.watch(clientSession.unwrapped()))

    override fun <T : Any> watch(clientSession: ClientSession, resultClass: Class<T>): ChangeStreamIterable<T> =
        SyncChangeStreamIterable(wrapped.watch(clientSession.unwrapped(), resultClass = resultClass))

    override fun watch(clientSession: ClientSession, pipeline: MutableList<out Bson>): ChangeStreamIterable<Document> =
        SyncChangeStreamIterable(wrapped.watch(clientSession.unwrapped(), pipeline))

    override fun <T : Any> watch(
        clientSession: ClientSession,
        pipeline: MutableList<out Bson>,
        resultClass: Class<T>
    ): ChangeStreamIterable<T> =
        SyncChangeStreamIterable(wrapped.watch(clientSession.unwrapped(), pipeline, resultClass))

    override fun getClusterDescription(): ClusterDescription = wrapped.getClusterDescription()

    private fun ClientSession.unwrapped() = (this as SyncClientSession).wrapped
}
